docs/architetture-software/README.md

```diff
@@ -0,0 +1,230 @@
+
+Questo tipo di _fitness function_ richiede solitamente più manutenzione rispetto alle _continuous fitness functions_, poiché deve essere adattato durante la vita del software, in base alle modifiche che vengono apportate.
+
+Alcuni esempi sono i test unitari, i test di integrazione, ecc.
+
+### Continuous Fitness functions
+
+Le _continuous fitness functions_ (funzioni di fitness _continue_) vengono eseguite in modo continuo sull'ambiente in analisi, tramite raccolta di metriche in tempo reale o comunque a cadenze rapide. Questo tipo di funzioni di fitness sono utili per monitorare il sistema e rilevare eventuali problemi prima che si verifichino, o analizzare dei trend.
+
+Questo tipo di funzioni ha da un lato il vantaggio di rappresentare la situazione reale, dall'altro lato ha il difetto di non essere replicabile, in quanto si basa su dati presi dall'ambiente che si sta analizzando (solitamente quello di produzione).
+
+Queste funzioni di fitness vengono spesso lanciate in un ambiente di produzione, in modo da avere contezza del reale contesto nel quale vengono eseguite.
+
+Alcuni esempi sono il monitoraggio dei tempi medi di risposta, il monitoraggio dei messaggi inviati e ricevuti in una coda, ecc.
+
+## Architecture Decision Records
+
+Gli _Architecture Decision Records_ (ADR) sono un modo per documentare le decisioni prese durante la progettazione di un sistema software. Questo tipo di documentazione è molto utile per capire il perché di determinate scelte e per capire come il sistema è stato progettato.
+
+A differenza di una classica documentazione tecnica o di un diagramma, le ADR sono molto più semplici da scrivere e da leggere, in quanto sono composte da pochi elementi:
+
+- **Titolo**: il titolo della decisione presa.
+- **Stato**: lo stato della decisione presa. Solitamente può essere _Proposta_, _Accettata_, _Completata_, _Rifiutata_ o _Sostituita_.
+- **Contesto**: il contesto in cui è stata presa la decisione.
+- **Decisione**: la decisione presa.
+- **Conseguenze**: le conseguenze della decisione presa.
+
+Le ADR vengono spesso utilizzate nelle aziende in cui si fa leva sul _BDD_ (Behavior Driven Development) per la stesura dei ticket nel proprio sistema di ticketing, puntando ad uno specifico ADR per dare informazioni aggiuntive e contesto a chi legge il ticket.
+
+## Quali sono le principali architetture software?
+
+Esistono diverse architetture software, ognuna con i suoi pro e i suoi contro. Quanto segue è un elenco non esaustivo delle principali architetture software, diventato però nel corso dei decenni uno standard de facto.
+
+Ad ogni architettura software è associato un elenco contenente i punti chiave che seguono:
+
+- **Start-up**: Il tempo necessario per implementare questa architettura software e cominciare a sviluppare funzionalità. Potremmo anche definirla come la _time to market_.
+- **Costi**: Il costo che si affronta nell'arco della vita di un progetto o un prodotto che implementi questa architettura.
```

Comment:
differenzierei i costi di implementazione dai costi di run: un conto è un progetto che costa poco lato implementativo ma un botto per il run (licenze e simili), un altro il viceversa.

Reply:
Volevo dare un'idea semplificata del concetto di _costo_ in quanto chi si occupa dello sviluppo solitamente ne è a conoscenza ma non influisce direttamente, tranne che in determinati contesti, mi sembrava eccessivo suddividere in due metriche. Come la vedi?